pages/en/lb4/Creating-components.md

```diff
@@ -307,3 +307,297 @@ class EmailComponent {
   }
 }
 ```
+
+## Creating your own servers 
+
+LoopBack 4 has the concept of a Server, which you can use to create your own
+implementations of REST, SOAP, gRPC, MQTT and more. For an overview, see 
+[Server](Server.html).
+
+### A Basic Example
+The concept of servers in LoopBack doesn't actually require listening on ports;
+you can use this pattern to create workers that process data, write logs, or
+whatever you'd like.
+
+Let's create an example server! Please note that this is not a complete example:
+
+```ts
+import {Server} from '@loopback/core';
+import {processReports} from '../util';
+export class ReportProcessingServer implements Server {
+  @inject('database') db: DefaultCrudRepository<Report, int>;
+  interval: NodeJS.Timer;
+  async start() {
+    // Run the report processor every 2 minutes.
+    let window = Date.now();
+    interval = setInterval(async () => {
+      const results = await db.find({
+        createdDate: {
+          $gt: window,
+        },
+      });
+      if (results && results.length > 0) {
+        results.forEach(async (rep) => {
+          await db.updateById(rep.id, await processReport(rep));
+        }); 
+      }
+      window = Date.now();
```

Comment:
```ts
const results = await db.find({
        createdDate: {
         $gt: window,
       },
     });
// ...
window = Date.now();
```

This creates a race condition. What if some records have been added after the database finished the query but before we have received all the matching records (and processed them)? Those will be silently ignored.

I understand you want to keep this example simple, but I don't think that's a valid excuse for teaching people to write subtly broken code.

Assuming `processReport` is idempotent (produces the same result when executed multiple times on the same record), the problem can be fixed by storing the current time _before_ querying the database.

```ts
class {
  // ...
  async processReports() {
    const since = this.window; // please, use a better name than "window"
    this.window = Date.now();

    const results = await db.find({
      createdDate: {
        $gt: since,
      },
    });
    // etc. (no changes there)
  }
}
```